src/track/globaltrackcache.cpp

```diff
@@ -419,177 +513,272 @@ GlobalTrackCacheResolver GlobalTrackCache::resolve(
+            TrackRefPtr(std::move(pTrack), std::move(trackRef)));
 }
 
-TrackRef GlobalTrackCache::updateTrackIdInternal(
-        const TrackPointer& pTrack,
+TrackRef GlobalTrackCache::initTrackIdInternal(
+        TrackPointer pTrack,
         const TrackRef& trackRef,
         TrackId trackId) {
+
+    DEBUG_ASSERT(!trackRef.getId().isValid());
     DEBUG_ASSERT(trackId.isValid());
-    if (trackRef.getId() != trackId) {
-        DEBUG_ASSERT(m_tracksById.end() == m_tracksById.find(trackId));
-        TrackRef trackRefWithId(trackRef, trackId);
-        Item item(trackRefWithId, pTrack);
-        m_tracksById.insert(
-                item.ref.getId(),
-                item);
-        m_tracksByCanonicalLocation.insert(
-                item.ref.getCanonicalLocation(),
-                item);
-        return trackRefWithId;
-    }
-    return trackRef;
+    TrackRef trackRefWithId(trackRef, trackId);
+
+    // Insert item by id
+    DEBUG_ASSERT(m_tracksById.find(trackId) == m_tracksById.end());
+    m_tracksById.insert(
+            trackId,
+            Item(trackRefWithId, std::move(pTrack)));
+
+    // Update item by track location
+    auto i = m_tracksByCanonicalLocation.find(
+            trackRef.getCanonicalLocation());
+    DEBUG_ASSERT(i != m_tracksByCanonicalLocation.end());
+    i.value().ref = trackRefWithId;
+
+    DEBUG_ASSERT(m_tracksById.find(trackId).value() ==
+            m_tracksByCanonicalLocation.find(trackRef.getCanonicalLocation()).value());
+    return trackRefWithId;
 }
 
-void GlobalTrackCache::evict(
-        Track* pTrack) {
-    DEBUG_ASSERT(pTrack != nullptr);
-    const auto trackRef = TrackRef::fromFileInfo(
-            pTrack->m_fileInfo,
-            pTrack->m_record.getId());
+void GlobalTrackCache::afterEvicted(
+        GlobalTrackCacheLocker* pCacheLocker,
+        Track* pEvictedTrack) {
+    DEBUG_ASSERT(pEvictedTrack);
+
+    // It can produce dangerous signal loops if the track is still
+    // sending signals while being saved! All references to this
+    // track have been dropped at this point, so there is no need
+    // to send any signals.
+    // See: https://bugs.launchpad.net/mixxx/+bug/136578
+    // NOTE(uklotzde, 2018-02-03): Simply disconnecting all receivers
+    // doesn't seem to work reliably. Emitting the clean() signal from
+    // a track that is about to deleted may cause access violations!!
+    pEvictedTrack->blockSignals(true);
+
+    // Keep the cache locked while evicting the track object!
+    // The callback is given the chance to unlock the cache
+    // after all operations that rely on managed track ownership
+    // have been done, e.g. exporting track metadata into a file.
+    m_pEvictor->afterEvictedTrackFromCache(
+            pCacheLocker,
+            pEvictedTrack);
+
+    // At this point the cache might have been unlocked by the
+    // callback!!
+}
+
+bool GlobalTrackCache::evictAndDelete(
+        Track* pTrack,
+        bool evictUnexpired) {
```

Comment:
bool evictUnexpired is always false. can we remove it?

Reply:
Yes. This is indeed a remainder from a previous version. Now the destruction invokes the internal method directly.